tests/contrib/django_util/test_decorators.py

```diff
@@ -0,0 +1,250 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+
+from django import http
+import django.conf
+from django.contrib.auth.models import AnonymousUser, User
+import mock
+from six.moves import http_client
+from six.moves import reload_module
+from six.moves.urllib import parse
+from tests.contrib.django_util import TestWithDjangoEnvironment
+
+import oauth2client.contrib.django_util
+from oauth2client.contrib.django_util import decorators
+
+
+class OAuth2EnabledDecoratorTest(TestWithDjangoEnvironment):
+
+    def setUp(self):
+        super(OAuth2EnabledDecoratorTest, self).setUp()
+        self.save_settings = copy.deepcopy(django.conf.settings)
+
+        if hasattr(django.conf.settings, 'GOOGLE_OAUTH2_STORAGE_MODEL'):
+            del django.conf.settings.GOOGLE_OAUTH2_STORAGE_MODEL
+
+        # OAuth2 Settings gets configured based on Django settings
+        # at import time, so in order for us to reload the settings
+        # we need to reload the module
+        reload_module(oauth2client.contrib.django_util)
+        self.user = User.objects.create_user(
+            username='bill', email='bill@example.com', password='hunter2')
+
+    def tearDown(self):
+        super(OAuth2EnabledDecoratorTest, self).tearDown()
+        django.conf.settings = copy.deepcopy(self.save_settings)
+
+    def test_no_credentials_without_credentials(self):
+        request = self.factory.get('/test')
+        request.session = self.session
+
+        @decorators.oauth_enabled
+        def test_view(request):
+            return http.HttpResponse("test")  # pragma: NO COVER
+
+        response = test_view(request)
+        self.assertIs(response.status_code, http_client.OK)
+        self.assertIsNotNone(request.oauth)
+        self.assertFalse(request.oauth.has_credentials())
+        self.assertIsNone(request.oauth.http)
+
+    @mock.patch('oauth2client.contrib.dictionary_storage.OAuth2Credentials')
+    def test_has_credentials_in_storage(self, OAuth2Credentials):
+        request = self.factory.get('/test')
+        request.session = mock.MagicMock()
+
+        credentials_mock = mock.Mock(
+            scopes=set(django.conf.settings.GOOGLE_OAUTH2_SCOPES))
+        credentials_mock.has_scopes.return_value = True
+        credentials_mock.invalid = False
+        credentials_mock.scopes = set([])
+        OAuth2Credentials.from_json.return_value = credentials_mock
+
+        @decorators.oauth_enabled
+        def test_view(request):
+            return http.HttpResponse('test')
+
+        response = test_view(request)
+        self.assertEqual(response.status_code, http_client.OK)
+        self.assertEqual(response.content, b'test')
+        self.assertTrue(request.oauth.has_credentials())
+        self.assertIsNotNone(request.oauth.http)
+        self.assertEqual(
```

Comment:
`assertSetEqual` here? Always use the most specific assertion method that you can (subject to the API of the system under test). Is it important that `request.oauth.scopes` be a `set`, or that it be a sequence, or that it simply have the same items as the expected value?